crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1084,12 +996,212 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
             }
         }
 
+        // TODO: it may be a bit expensive to run it on each add, try to group.
+        // TODO: move it to `handle_event`?
+        self.maybe_adjust_date_dividers();
+
         // If we don't have a read marker item, look if we need to add one now.
         if !self.meta.has_up_to_date_read_marker_item {
             self.meta.update_read_marker(self.items);
         }
     }
 
+    /// Ensures that date separators are properly inserted/removed when needs
+    /// be.
+    fn maybe_adjust_date_dividers(&mut self) {
+        // We're going to record operations like inserting, replacing and removing day
+        // dividers. Since we may remove or insert new items, recorded offsets
+        // will need to be updated, and the list of items must be iterated in
+        // non-decreasing order.
+
+        #[derive(Debug)]
+        enum Operation {
+            Insert(usize, MilliSecondsSinceUnixEpoch),
+            Replace(usize, MilliSecondsSinceUnixEpoch),
+            Remove(usize),
+        }
+
+        let mut ops = vec![];
+
+        let mut prev_item: Option<(&Arc<TimelineItem>, MilliSecondsSinceUnixEpoch)> = None;
+        let mut last_event_ts = None;
+
+        for (i, item) in self.items.iter().enumerate() {
+            if let TimelineItemKind::Virtual(VirtualTimelineItem::DayDivider(ts)) = item.kind() {
+                // It's a day divider.
+
+                if let Some((prev_item, prev_ts)) = prev_item {
+                    if prev_item.is_day_divider() {
+                        trace!("removing duplicate day divider @ {i}");
+                        // This day divider is preceded by another one: remove the current one.
+                        ops.push(Operation::Remove(i));
+                    } else if prev_item.is_event() {
+                        // This day divider is preceded by an event.
+                        if timestamp_to_date(prev_ts) == timestamp_to_date(*ts) {
+                            // The event has the same date as the day divider: remove the day
+                            // divider.
+                            trace!(
+                                "removing day divider following event with same timestamp @ {i}"
+                            );
+                            ops.push(Operation::Remove(i));
+                        }
+                    }
+                } else {
+                    // No interesting item prior to the day divider: it must be
+                    // the first one, nothing to do.
+                }
+
+                prev_item = Some((item, *ts));
+            }
+
+            if let Some(event) = item.as_event() {
```

Comment:
/me whispers *because match cause an extra level of indent :pleading_face:*

We could definitely use a match here and below, which also makes handling the other virtual items more explicit, so I'll do it, thanks.